Add a plugin design doc

bamarni opened the pull request:
This adds the plugin design doc from:

https://docs.google.com/document/d/1UN3rPhmtdpy2dz1jn3XobTJlOxQtXujbW3ExvBkmi14/edit

It removes some background section to fit more into the style of the
design pages in this repo, and updates some sections to reflect some
recent changes that were made.

The second commit is a design update to the plugin system in order to pass
additionnal env variables to plugins whenever a cluster is configured.

The main goal being to make it easier for plugins to interact with
DC/OS, as they currently have to read the config TOML file by themselves
or invoke back the `dcos` binary.

br-lewis commented on design/plugin.md:
```diff
@@ -0,0 +1,131 @@
+
+For example, if a `dcos-hello` executable in a plugin has a `hello` top level command, it would be called like `env/bin/dcos-hello hello [...]`.
+
+Executables contained within plugins are called synchronously from the CLI by spawning a new child process. The CLI waits for the child process to complete. The CLI makes available its own executable path to the child via an ENV variable `DCOS_CLI_EXECUTABLE_PATH`.
```
The description of `DCOS_CLI_EXECUTABLE_PATH` probably belongs in the list below.

bamarni replied:
The list below is for env variables which are only set when a cluster is attached.

br-lewis replied:
Ah I see now.

br-lewis commented on design/setup.md:
```diff
@@ -102,3 +103,29 @@ and the CLI is attached to it. A cluster config looks like this:
+package, service, task.
+
+When the core plugin can't be installed (eg. insufficient Cosmos permission or air-gapped environment),
+it then fallbacks to installing it from the DC/OS CLI binary itself, which bundles a core plugin.
```
"fallbacks" -> "falls back".

We should make clear that which version of the core CLI it has bundled will be dependent on the version of the wrapper CLI you happen to be using and that it won't check the version of the core it has against the DC/OS cluster.

bamarni commented:
I've addressed the comments, I mainly wanted to add the second commit here as it's needed for https://jira.mesosphere.com/browse/DCOS_OSS-4101 however we didn't have the design for the current system yet.
